parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -719,32 +719,62 @@ modifyAtM path f b = case Path.uncons path of
     -- step the branch by updating its children according to fixup
     pure $ step (setChildBranch seg child') b
 
-stepAt0 :: Applicative m => Path
-                         -> (Branch0 m -> Branch0 m)
-                         -> Branch0 m -> Branch0 m
-stepAt0 p f = runIdentity . stepAt0M p (pure . f)
-
 -- stepManyAt0 consolidates several changes into a single step
-stepManyAt0 :: (Applicative m, Foldable f)
+stepManyAt0 :: forall f m . (Applicative m, Foldable f)
            => f (Path, Branch0 m -> Branch0 m)
            -> Branch0 m -> Branch0 m
-stepManyAt0 actions b = foldl' (\b (p, f) -> stepAt0 p f b) b actions
-
-stepManyAt0M :: (Monad m, Monad n, Foldable f)
+stepManyAt0 actions b = go (toList actions) b where
+  go :: [(Path, Branch0 m -> Branch0 m)] -> Branch0 m -> Branch0 m
+  go actions b = let
+    -- combines the functions that apply to this level of the tree
+    currentAction b = foldl' (\b f -> f b) b [ f | (Path.Empty, f) <- actions ]
+
+    -- groups the actions based on the child they apply to
+    childActions :: Map NameSegment [(Path, Branch0 m -> Branch0 m)]
+    childActions =
+      List.multimap [ (seg, (rest,f)) | (seg :< rest, f) <- actions ]
+
+    -- alters the children of `b` based on the `childActions` map
+    stepChildren :: Map NameSegment (Branch m) -> Map NameSegment (Branch m)
+    stepChildren m0 = foldl' g Map.empty (Map.keysSet m0 <> Map.keysSet childActions)
+      where
+      -- this skips adding any empty children, which could happen depending
+      -- on the action functions provided (think: `b -> Branch.empty0`)
+      insert seg child m = if isEmpty child then m else Map.insert seg child m
+      g m seg = insert seg child m where
+        child = case Map.lookup seg childActions of
+          Just actions ->
+            -- this `findWithDefault` is important, allows the stepManyAt
+            -- to create new children at deeper paths than exist in the
+            -- original Branch
+            step (go actions) (Map.findWithDefault empty seg m0)
```

Comment:
`(go actions)` aggregates all of the actions at this level and below into a single action. :+1: